src/Root.cpp

```diff
@@ -332,6 +332,24 @@ void cRoot::Start(std::unique_ptr<cSettingsRepositoryInterface> a_OverridesRepo)
 
 void cRoot::StopServer()
 {
+	// Kick all players from the server with custom disconnect message
+	class cPlayerCallback : public cPlayerListCallback
+	{
+		AString m_ShutdownMessage;
+		virtual bool Item(cPlayer * a_Player)
+		{
+			a_Player->GetClientHandlePtr()->Kick(m_ShutdownMessage);
+			return false;
+		}
+	public:
+		cPlayerCallback(AString a_ShutdownMessage) : m_ShutdownMessage(a_ShutdownMessage) {}
+	};
+	if (m_Server->GetNumPlayers())
```

Comment:
I think the flow should be inverse - set a flag in `cPlayerCallback` that a D/C message has been sent to a player, execute it `ForEachPlayer` and then check the value:
```cpp
cRoot::Get()->ForEachPlayer(callback);
if (callback.m_HasSentDisconnect)
{
  std::this_thread::sleep_for(...);
}
```

As it is now, the code is prone to race conditions - a player can connect or disconnect between the `GetNumPlayers()` call and the `ForEachPlayer()` call. Not that it would matter much here, but let's not get in the habit of having problematic code.